src/Stack/Build/Execute.hs

```diff
@@ -1809,22 +1810,39 @@ singleTest topts testsToRun ac ee task installedMap = do
                 tixPath <- liftM (pkgDir </>) $ parseRelFile $ exeName ++ ".tix"
                 exePath <- liftM (buildDir </>) $ parseRelFile $ "build/" ++ testName' ++ "/" ++ exeName
                 exists <- doesFileExist exePath
+                -- doctest relies on template-haskell in QuickCheck-based tests
+                thGhcId <- case find ((== "template-haskell") . pkgName . dpPackageIdent. snd)
+                                (Map.toList $ eeGlobalDumpPkgs ee) of
+                  Just (ghcId, _) -> return ghcId
+                  Nothing -> error "template-haskell is a wired-in GHC boot library but it wasn't found"
                 packageIds <- forMaybeM (M.toList $ packageDeps package) $ \(name, dv) -> do
                     let pkgId = PackageIdentifier name nullVersion
                     case Map.lookupGT pkgId allDepsMap of
                         Just (PackageIdentifier name' version, ghcPkgId)
                             | name' == name && dvType dv == AsLibrary &&
                               version `withinRange` dvVersionRange dv ->
-                            return $ Just (unGhcPkgId ghcPkgId)
+                            return $ Just ghcPkgId
                         _ -> do
                             logWarn $ "Could not find GHC package id for dependency " <> fromString (packageNameString name)
                             return Nothing
-                -- env var HASKELL_PACKAGE_IDS is used by doctest so module names for
+                -- env variable GHC_ENVIRONMENT is set for doctest so module names for
                 -- packages with proper dependencies should no longer get ambiguous
                 -- see e.g. https://github.com/doctest/issues/119
-                let usePackageIds pc = modifyEnvVars pc $ \envVars ->
-                      Map.insert "HASKELL_PACKAGE_IDS" (T.unwords packageIds) envVars
-                menv <- liftIO $ usePackageIds =<< configProcessContextSettings config EnvSettings
+                let setEnv f pc = modifyEnvVars pc $ \envVars ->
+                      Map.insert "GHC_ENVIRONMENT" (T.pack f) envVars
+                    fp = toFilePath $ eeTempDir ee </> $(mkRelFile "test-ghc-env")
+                    snapDBPath = toFilePathNoTrailingSep (bcoSnapDB $ eeBaseConfigOpts ee)
+                    localDBPath = toFilePathNoTrailingSep (bcoLocalDB $ eeBaseConfigOpts ee)
+                    ghcEnv = mconcat $
+                        "clear-package-db\n" :
```

Comment:
It's slightly more efficient to bypass the `mconcat` and do `"foo" <> "bar" <> ... <> foldMap f rest`, but that's not terribly important here.

Reply:
I'll fix it anyway

Reply:
Done in 99059212